_source/_posts/2022-06-01-cloud-native-java-microservices-with-istio.md

```diff
@@ -0,0 +1,604 @@
+
+{% youtube zGpnIhRgMaM %}
+
+**Pre-requisites**
+
+- A [Google Cloud Platform](https://cloud.google.com/) account
+- [Docker](https://www.docker.com/get-started) installed on your machine
+- [Node.js](https://nodejs.org/en/) installed on your machine
+- [JHipster installed](https://www.jhipster.tech/installation/) on your machine
+- [Google Cloud SDK](https://cloud.google.com/sdk/docs/install) installed and configured on your machine
+- [kubectl](https://kubernetes.io/docs/tasks/tools/) or [KDash](https://github.com/kdash-rs/kdash)
+- Basic understanding of Java, Spring, Containers, and Kubernetes
+
+## Why Build Cloud-Native Microservices Using a Service Mesh?
+
+Before we dive into building a cloud-native microservice stack, let's look at what a service mesh is and the benefits of using one.
+
+A service mesh provides features to help with common distributed microservice challenges. Like service discovery, routing, load balancing, and so on. Today we will be using [Istio](https://istio.io/), one of the most popular service mesh solutions available. Istio is tailored for distributed application architectures, especially those you might run in Kubernetes. Istio plays extremely nice with Kubernetes, so nice that you might think that it's part of the Kubernetes platform itself. Istio isn't the only service mesh around; we also have platforms like [Linkerd](https://linkerd.io/) and [Consul](https://www.consul.io/) which are also quite popular.
+
+Istio specifically provides the following features.
+
+- Secure service-to-service communication over TLS. Of course, with support for identity-based authentication and authorization.
+- Service discovery, so that your microservices can discover each other
+- Automatic load balancing for the services
+- Traffic control features like routing, circuit breaking, retries, fail-overs, and fault injection.
+- A pluggable policy layer that can enforce stuff like access control, rate limiting, A/B testing, traffic splits, quotas, etc.
+- It also provides automatic metrics, logs, and traces for all traffic within the cluster from Ingres to Egress and between pods.
+
+### What is Istio Service Mesh?
+
+Let's take a quick look at Istio internals. The Istio architecture can be classified into two distinct planes.
+
+{% img blog/cloud-native-java-microservices-with-istio/istio-architecture.png alt:"Istio Service Mesh Architecture" width:"900" %}{: .center-image }
+
+**Data plane**: It's made of [Envoy](https://www.envoyproxy.io/) proxies deployed as sidecars to our application containers. Envoy is a high-performance, lightweight distributed proxy. It controls all the incoming and outgoing traffic to the container it is attached to.
+
+**Control plane**: It consists of the istiod demon, and it manages and configures the envoy proxies to route traffic. It also enforces policies and collects telemetry. It has components like Pilot for traffic management, Citadel to manage security, and Galley to manage configurations.
```

Comment:
You might want to order these planes in the same order as they are in the diagram above.